File: LUA/Problem004 - Largest palindrome product.lua
--[[
A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.
]]

-- Largest palindrome we have found so far
largest = 1

-- Start out first number at 100, the smallest 3-digit number
for i = 100, 999 do
	-- Start our second number at wherever our first is at, everything below that is duplicate
	for j = i, 999 do
		product = i * j
		-- First check if the product is the largest so far, which is faster to check than if it is a palindrome
		if product > largest then
			-- If it is the largest, reverse the number so we can check if it is a palindrome. If it is, remember it as the largest so far.
			palindrome = tonumber(string.reverse(tostring(product)))
			if palindrome == product then
				largest = product
			end
		end
	end
end

io.write("The largest palindrome product of two 3-digit numbers is ", largest)

--[[

Below is an alternative method of reversing the number, without casting to string and back.

temp = product
palindrome = 0
while temp > 0 do
palindrome = palindrome * 10
palindrome = palindrome + (temp % 10)
temp = math.floor(temp / 10)
end
if palindrome == product then
	largest = product
end
]]

File: Problem 004 - Largest palindrome product.cpp
// Find largest palindrome product of two 3-digit numbers.
#include <iostream>

using namespace std;

int main() {
    int largest = 0;
    // Our two numbers range from 100 to 999, since those are the only 3-digit numbers.
    for (int i = 100; i < 1000; i++){
        for (int j = 100; j < 1000; j++){
            int product = i*j;
            // Check if the product is the largest we've found so far first.
            if (product > largest) {
                // Now we reverse the product to see if it is a palindrome.
                int reverse = 0;
                int temp = product;
                while (temp != 0){
                    // Remainder division by 10 leaves only the last digit.
                    int digit = temp % 10;
                    reverse *= 10;
                    reverse += digit;
                    temp -= digit;
                    temp /= 10;
                }
                // If our product and its reverse are the same it's a palindrome.
                if (product == reverse) {
                    largest = product;
                }
            }
        }
    }
    cout<<"Largest palindrome product of two 3-digit numbers is: "<<largest;
    return 0;
}
